### PR TITLE
Add datahub crawler

### PR DESCRIPTION
**Pull request by usefulalgorithm**
### 🤔 Why?

Adds datahub crawler. Retrieves the dataset tags, table descriptions, and ownership assignments.
Column description is not available, datahub api does not return them in their responses.

### 🤓 What?

- Added datahub crawler.
- Added unit test.

### 🧪 Tested?

Added unit test.
End to end testing:
截圖 2023-12-21 凌晨3 02 43

### ☑️ Checks

- [x] My PR contains actual code changes, and I have updated the version number in `pyproject.toml`.

### CODE REVIEW

File: metaphor/datahub/gql_parser.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+            platform=meta_platform,
+        )
+
+    @property
+    def description_assignment(self) -> Optional[DescriptionAssignment]:
+        asset_descriptions = None
+        if self.properties and self.properties.description:
+            asset_descriptions = [AssetDescription(self.properties.description)]
+
+        # XXX: Datahub bug, graphql api does not return any column description.
```
Comment (usefulalgorithm) [inline metaphor/datahub/gql_parser.py:238]: ORZ

File: metaphor/datahub/gql_parser.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        meta_platform = DATAHUB_PLATFORM_MAPPING.get(
+            self.platform.name, MetaDataPlatform.UNKNOWN
+        )
+        if meta_platform is MetaDataPlatform.UNKNOWN:
+            logger.warning(
+                f"Found unknown data platform for dataset {name}: {self.platform.name}"
+            )
+
+        return DatasetLogicalID(
+            account=None,  # FIXME This is not stored in datahub, have to find another way to get this
```
Comment (usefulalgorithm) [inline metaphor/datahub/gql_parser.py:227]: what to do with this?

Comment (usefulalgorithm) [inline metaphor/datahub/gql_parser.py:227]: what to do with this?

Comment (usefulalgorithm) [inline metaphor/datahub/gql_parser.py:227]: what to do here...

File: metaphor/datahub/gql_parser.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        return self.tags.system_tags
+
+    def has_additional_information(self) -> bool:
+        return (
+            self.ownership_assignment is not None
+            or self.description_assignment is not None
+            or self.system_tags is not None
+        )
+
+    def as_meta_dataset(self, config: DatahubConfig) -> MetaDataset:
```
Comment (elic-eon) [inline metaphor/datahub/gql_parser.py:281]: `as_metaphor_dataset`?

File: metaphor/datahub/gql_parser.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+        # It's possible that we want to split the name by the platform delimiters to get part names.
+        name = normalize_full_dataset_name(self.name)
+
+        meta_platform = DATAHUB_PLATFORM_MAPPING.get(
+            self.platform.name, MetaphorDataPlatform.UNKNOWN
+        )
+        if meta_platform is MetaphorDataPlatform.UNKNOWN:
+            logger.warning(
+                f"Found unknown data platform {self.platform.name}, will not ingest dataset {name}"
+            )
```
Comment (elic-eon) [inline metaphor/datahub/gql_parser.py:223]: So we will generate the entity but the ingestion will skip for `UNKNOWN` platform?

File: metaphor/datahub/gql_parser.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+    tags: Optional[GlobalTags]
+    ownership: Optional[Ownership]
+    schemaMetadata: Optional[SchemaMetadata]
+    editableSchemaMetadata: Optional[EditableSchemaMetadata]
+
+    def get_logical_id(self, config: DatahubConfig) -> DatasetLogicalID:
+        # It's possible that we want to split the name by the platform delimiters to get part names.
+        name = normalize_full_dataset_name(self.name)
+
+        meta_platform = DATAHUB_PLATFORM_MAPPING.get(
```
Comment (elic-eon) [inline metaphor/datahub/gql_parser.py:217]:
```suggestion
        metaphor_platform = DATAHUB_PLATFORM_MAPPING.get(
```

File: metaphor/datahub/gql_parser.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+            self.platform.name, MetaphorDataPlatform.UNKNOWN
+        )
+        if meta_platform is MetaphorDataPlatform.UNKNOWN:
```
Comment (elic-eon) [inline metaphor/datahub/gql_parser.py:220]:
```suggestion
        if metaphor_platform is MetaphorDataPlatform.UNKNOWN:
```

**Review by elic-eon — COMMENTED** [review]
Otherwise, LGTM

File: metaphor/datahub/gql_parser.py (hunk 4 above)
Comment (usefulalgorithm) [reply metaphor/datahub/gql_parser.py:223]: do you think it's better to have this as an external platform?

Comment (usefulalgorithm) [reply metaphor/datahub/gql_parser.py:223]: er... yeah it's probably better to have this as an external platform. will change that.

Comment (elic-eon) [reply metaphor/datahub/gql_parser.py]: Maybe it's better not to generate these datasets. wdyt @mars-lan

Comment (mars-lan) [reply metaphor/datahub/gql_parser.py]: It's probably safe to ignore any unmatched platform for now. We won't have the correct UI to show them anyway.

**Review by elic-eon — DISMISSED** [review]
LGTM Great work.

File: pyproject.toml [hunk 7]
```diff
@@ -21,19 +21,20 @@ aws-assume-role-lib = "^2.10.0"
...
-botocore = "^1.31.57"
+botocore = "^1.31.58"
 canonicaljson = "^2.0.0"
 confluent-kafka = { version = "^2.3.0", optional = true }
 databricks-sdk = { version = "^0.14.0", optional = true }
 databricks-sql-connector = { version = "^3.0.0", optional = true }
 GitPython = "^3.1.37"
 google-cloud-bigquery = { version = "^3.1.0", optional = true }
 google-cloud-logging = { version = "^3.5.0", optional = true }
+gql = { extras = ["all"], version = "^3.4.1", optional = true }
```
Comment (mars-lan) [inline pyproject.toml:32]: Do we really need `all` extras? https://gql.readthedocs.io/en/latest/intro.html#less-dependencies

Comment (usefulalgorithm) [reply pyproject.toml]: good catch, thx

**Review by elic-eon — APPROVED** [review]

**Pull request merged**